cc/algorithms/partition-selection-strategies_test.cc

```diff
@@ -0,0 +1,373 @@
+  									.SetMaxPartitionsContributed(3).Build();
+  	EXPECT_THAT(failed_build.status().code(),
+                Eq(base::StatusCode::kInvalidArgument));
+   	std::string message(std::string(failed_build.status().message()));
+  	EXPECT_THAT(message, MatchesRegex("^Delta has to be finite.*"));
+}
+
+TEST(PartitionSelectionStrategiesTest, PreaggPartitionSelectionInvalidDelta) {
+	PreaggPartitionSelection::Builder test_builder;
+  	auto failed_build = test_builder.SetEpsilon(4.5).SetDelta(6.0)
+  									.SetMaxPartitionsContributed(7).Build();
+  	EXPECT_THAT(failed_build.status().code(),
+                Eq(base::StatusCode::kInvalidArgument));
+   	std::string message(std::string(failed_build.status().message()));
+  	EXPECT_THAT(message, MatchesRegex("^Delta has to be in the interval.*"));
+}
+
+TEST(PartitionSelectionStrategiesTest,
+	 PreaggPartitionSelectionUnsetMaxPartitionsContributed) {
+	PreaggPartitionSelection::Builder test_builder;
+  	auto failed_build = test_builder.SetEpsilon(0.8).SetDelta(0.9).Build();
+  	EXPECT_THAT(failed_build.status().code(),
+                Eq(base::StatusCode::kInvalidArgument));
+   	std::string message(std::string(failed_build.status().message()));
+  	EXPECT_THAT(message, MatchesRegex("^Max number of partitions a user can"
+  									  " contribute to has to be set.*"));
+}
+
+TEST(PartitionSelectionStrategiesTest,
+	 PreaggPartitionSelectionNegativeMaxPartitionsContributed) {
+	PreaggPartitionSelection::Builder test_builder;
+  	auto failed_build = test_builder.SetEpsilon(0.1).SetDelta(0.2)
+  									.SetMaxPartitionsContributed(-3).Build();
+  	EXPECT_THAT(failed_build.status().code(),
+                Eq(base::StatusCode::kInvalidArgument));
+   	std::string message(std::string(failed_build.status().message()));
+  	EXPECT_THAT(message, MatchesRegex("^Max number of partitions a user can"
+  									  " contribute to has to be positive.*"));
+}
+
+//We expect the probability of keeping a partition with one user
+//will be approximately delta
+TEST(PartitionSelectionStrategiesTest, PreaggPartitionSelectionOneUser) {
+	PreaggPartitionSelection::Builder test_builder;
+	std::unique_ptr<PartitionSelectionStrategy> build =
+		test_builder.SetEpsilon(0.5).SetDelta(0.02)
+					.SetMaxPartitionsContributed(1).Build().ValueOrDie();
+  	double num_kept = 0.0;
+  	for(int i = 0; i < 1000000; i++) {
+  		if(build->ShouldKeep(1))
+  			num_kept++;
+  	}
+  	EXPECT_THAT(num_kept/1000000, DoubleNear(build->GetDelta(), 0.001));
+ }
+
+//We expect the probability of keeping a partition
+//with no users will always be zero
+TEST(PartitionSelectionStrategiesTest, PreaggPartitionSelectionNoUsers) {
+	PreaggPartitionSelection::Builder test_builder;
+	std::unique_ptr<PartitionSelectionStrategy> build =
+		test_builder.SetEpsilon(0.5).SetDelta(0.02)
+					.SetMaxPartitionsContributed(1).Build().ValueOrDie();
+  	double num_kept = 0.0;
+  	for(int i = 0; i < 100000; i++) {
```

Comment:
Rather than calculating the percentage kept, just EXPECT_FALSE every time. You can probably reduce the number of samples by a lot there too.

Reply:
I just uh... did one sample for this since the whole test suite is meant to be run many times. Also that made sense with the EXPECT_FALSE to me?  O__o I hope that's sufficient...

Reply:
I'd still use more than one sample. If this fails, there's a decent change that it happens only sometimes. We don't want a test that fails sometimes and runs sometimes (that counts as flaky), so multiple samples is still a good idea.